runtime/src/iree-amd-aie/driver/xrt/direct_command_buffer.cc

```diff
@@ -0,0 +1,446 @@
+        set, binding_count, IREE_HAL_XRT_MAX_DESCRIPTOR_SET_BINDING_COUNT);
+  }
+
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  xrt::bo** current_bindings = command_buffer->descriptor_sets[set].bindings;
+  for (iree_host_size_t i = 0; i < binding_count; i++) {
+    const iree_hal_descriptor_set_binding_t* binding = &bindings[i];
+    if (binding->buffer) {
+      continue;
+    }
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "unimplemented null buffer in push descriptor set");
+  }
+  for (iree_host_size_t i = 0; i < binding_count; i++) {
+    const iree_hal_descriptor_set_binding_t* binding = &bindings[i];
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                         &binding->buffer));
+    std::unique_ptr<xrt::bo> sub_buffer;
+    try {
+      xrt::bo* device_buffer = iree_hal_xrt_buffer_handle(
+          iree_hal_buffer_allocated_buffer(binding->buffer));
+      iree_device_size_t offset = iree_hal_buffer_byte_offset(binding->buffer);
+      iree_device_size_t length = binding->length;
+      // allocate a sub buffer of the device buffer as the binding object
+      sub_buffer = std::make_unique<xrt::bo>(*device_buffer, length,
+                                             offset + binding->offset);
+      current_bindings[binding->binding] = sub_buffer.release();
+    } catch (...) {
+      IREE_TRACE_ZONE_END(z0);
+      return iree_make_status(
+          IREE_STATUS_RESOURCE_EXHAUSTED,
+          "could not allocate memory for sub buffers being binded");
+    }
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_dispatch(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_executable_t* executable, int32_t entry_point,
+    uint32_t workgroup_x, uint32_t workgroup_y, uint32_t workgroup_z) {
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+  // Lookup kernel parameters used for side-channeling additional launch
+  // information from the compiler.
+  iree_hal_xrt_kernel_params_t kernel_params;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_xrt_native_executable_entry_point_kernel_params(
+              executable, entry_point, &kernel_params));
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                       &executable));
+  xrt::kernel kernel = *kernel_params.kernel;
+  xrt::bo instr = *kernel_params.instr;
+  uint32_t num_instr = kernel_params.num_instr;
+
+  xrt::run run = xrt::run(kernel);
+
+  // index to push arguments on the kernel;
```

Comment:
Let's be consistent regarding comments -- Captialize the first char, end with `.`.